src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -129,7 +129,47 @@ public static ODataItemBase ReadResourceOrResourceSet(this ODataReader reader)
                         }
 
                         break;
+                    case ODataReaderState.DeltaResourceSetStart:
+                        ODataDeltaResourceSet deltaResourceSet = (ODataDeltaResourceSet)reader.Item;
+                        Contract.Assert(deltaResourceSet != null, "ResourceSet should never be null.");
+
+                        ODataDeltaResourceSetWrapper deltaResourceSetWrapper = new ODataDeltaResourceSetWrapper(deltaResourceSet);
+
+                        AddNestedResourceInfo(ref topLevelItem, itemsStack, deltaResourceSetWrapper);
+
+                        itemsStack.Push(deltaResourceSetWrapper);
+                        break;
+
+                    case ODataReaderState.DeltaLink:
+                        ODataDeltaLink deltaLink = (ODataDeltaLink)reader.Item;
+                        Contract.Assert(deltaLink != null, "Delta link should never be null.");
+
+                        ODataDeltaLinkWrapper deltaLinkWrapper = new ODataDeltaLinkWrapper(deltaLink);
+
+                        Contract.Assert(itemsStack.Count > 0, "Delta link should never be reported as top-level item.");
+                        
+                        ODataDeltaResourceSetWrapper linkResourceSetWrapper = (ODataDeltaResourceSetWrapper)itemsStack.Peek();
+
+                        Contract.Assert(linkResourceSetWrapper != null, "ODataDeltaResourceSetWrapper for delta link should not be null.");
+
+                        linkResourceSetWrapper.Resources.Add(deltaLinkWrapper);
+                        
+                        break;
+                    case ODataReaderState.DeltaDeletedLink:
+                        ODataDeltaDeletedLink deltaDeletedLink = (ODataDeltaDeletedLink)reader.Item;
+                        Contract.Assert(deltaDeletedLink != null, "Delta deleted link should never be null.");
+
+                        ODataDeltaLinkWrapper deltaDeletedLinkWrapper = new ODataDeltaLinkWrapper(deltaDeletedLink);
+
+                        Contract.Assert(itemsStack.Count > 0, "Delta deleted link should never be reported as top-level item.");
+                        
+                        ODataDeltaResourceSetWrapper deletedLinkResourceSetWrapper = (ODataDeltaResourceSetWrapper)itemsStack.Peek();
+
+                        Contract.Assert(deletedLinkResourceSetWrapper != null, "ODataDeltaResourceSetWrapper for deleted delta link should not be null.");
 
+                        deletedLinkResourceSetWrapper.Resources.Add(deltaDeletedLinkWrapper);
```

Comment:
It seems we can combine  "ODataReaderState.DeltaLink" & "ODataReaderState.DeltaDeletedLink"

Reply:
I was thinking the same thing -- they are *almost* the same.  @ did you figure out how to combine (or move common implementation to a shared method)?

---
In reply to: [512320557](https://github.com/OData/WebApi/pull/2275#discussion_r512320557) [](ancestors = 512320557)